src/odgp.rs

```diff
@@ -14,10 +14,14 @@ const LAMBDA: ZRootTwo = ZRootTwo {
     b: IBig::ONE,
 };
 
-pub fn solve_odgp(i: Interval, j: Interval) -> Vec<ZRootTwo> {
-    if i.width() < ib_to_bf_prec(IBig::ZERO) || j.width() < ib_to_bf_prec(IBig::ZERO) {
-        return vec![];
-    }
+pub fn solve_odgp(i: Interval, j: Interval) -> impl Iterator <Item = ZRootTwo> {
+
+    // Can't return two different iterator types. So we can't do this check.
+    // I checked with dbg! to confirm that omitting this check is ok:
+    // If the condition is true, then an empty iterator is returned at the end.
+    //    if i.width() < ib_to_bf_prec(IBig::ZERO) || j.width() < ib_to_bf_prec(IBig::ZERO) {
+    //        return vec![].into_iter();
+    //    }
```

Comment:
Technically you can solve this problem by wrapping the output in `Box`:

```rust
pub fn solve_odgp(i: Interval, j: Interval) -> Box<dyn Iterator <Item = ZRootTwo>> {

    if i.width() < ib_to_bf_prec(IBig::ZERO) || j.width() < ib_to_bf_prec(IBig::ZERO) {
        return Box::new(vec![].into_iter());
    }

    return Box::new(...)
}
```
I am fine with whatever version you prefer.

Reply:
Yes. Many of these iterators are passed in and out of functions that create more iterators. If all the concrete types are known, the compiler should have opportunities for significant optimization. I'm not 100% sure about this particular case, however. In any case, the `Box` would kill this opportunity.

But the point is moot. I alluded to it in the comment, but was not very explicit: I checked that if the condition is satisfied the result is an empty iterator, which is what you want. (At the moment, I can't recall exactly how thorough my test was. i.e. did I test it in the larger context.)

Reply:
Mostly for the exercise, I added a couple of tests for this case (an empty interval is input). One tests that the returned iterator is empty. The next tests that a caller (which constructs a Vec) gets the empty iterator and returns an empty `Vec`.